cloudbuild/python-scripts/parse_logs.py

```
@@ -0,0 +1,281 @@
+    client = bigquery.Client(project=client_project_name)
+    table = client.get_table(table=table_ref)
+    row_count = table.num_rows
+    return row_count
+
+def extract_metric(logs_as_string, metric_string, delimiter):
+    total_metric_sum_in_blob = 0
+    # Keep on finding the metric value as there can be
+    # 1 or more outputs in a log file.
+    metric_pattern = r'{}\s*(.*?)\s*{}'.format(re.escape(metric_string), re.escape(delimiter))
+    metric_pattern = re.compile(metric_pattern)
+    matches = metric_pattern.finditer(logs_as_string)
+    records_read = [int(record_read.group(1).strip()) for record_read in matches]
+    total_metric_sum_in_blob += sum(records_read)
+    return total_metric_sum_in_blob
+
+def check_query_correctness(logs_as_string):
+    query_records_pattern = r'\[\s(.*),\s(.*)\s\]'
+    # Find all matches of the pattern in the string
+    matches = re.findall(query_records_pattern, logs_as_string)
+    # Extract and print all pairs of HOUR and DAY
+    if matches:
+        for match in matches:
+            hour = match[0].strip()
+            day = match[1].strip()
+            if hour !='17' or day !='17':
+                raise RuntimeError(f'[Log: parse_logs ERROR] Incorrect query result obtained!')
+    else:
+        # If no such matches are found.
+        print(f'[Log: parse_logs INFO] No query result obtained in this file')
+        return False
+    return True
+
+def get_blob_and_check_metric(
+    gcs_log_object, cluster_temp_bucket, metric_string, metric_delimiter, query
+):
+    logs_as_string = ''
+    try:
+        storage_client = storage.Client()
+        bucket = storage_client.get_bucket(cluster_temp_bucket)
+        blob = bucket.blob(gcs_log_object)
+        logs_as_string = blob.download_as_text(encoding='latin-1')
+    except Exception as e:
+        print(f'[Log: parse_logs WARNING] File {gcs_log_object} Not Found.\nError: {e}')
```

We should raise an exception or return -1 here

---

This is not an exception, there are cases (typically when number of workers increases) when all workers do not produce log files. In that case we just look at files obtained via other workers. We are able to geet all the records as well as a correct count of the total records read by looking at the leftover logs.